File: src/studio/program/element/And.java
package studio.program.element;

import studio.program.Program;
import studio.program.Signal;

public class And extends Block {
    public And() {
        super();
        text = "AND";
        name = "and";
    }

    @Override
    public void tick(double dt) {

    }

    @Override
    public void createPins(Program program) {
        Pin p0 = new Pin(this, Signal.Type.DISCRETE, Pin.Flow.INPUT);
        p0.setSide(Pin.Side.LEFT);
        p0.setAttachmentPoint(-width / 2, -20);

        Pin p1 = new Pin(this, Signal.Type.DISCRETE, Pin.Flow.INPUT);
        p1.setSide(Pin.Side.LEFT);
        p1.setAttachmentPoint(-width / 2, 20);

        Pin p2 = new Pin(this, Signal.Type.DISCRETE, Pin.Flow.OUTPUT);
        p2.setSide(Pin.Side.RIGHT);
        p2.setAttachmentPoint(width / 2, 0);

        program.addElement(p0);
        program.addElement(p1);
        program.addElement(p2);

        addPin(p0);
        addPin(p1);
        addPin(p2);
    }
}

File: src/studio/blocklib/definitions/And.kt
package studio.blocklib.definitions

import studio.blocklib.BlockDefinition
import studio.program.Block
import studio.program.Pin
import studio.program.Variable
import studio.shape.Rectangle

class And : BlockDefinition() {
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // CONSTANTS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    companion object {
        val NAME: String = "and"
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // CONSTRUCTOR
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    init {
        // I literally just want to use kotlin for the string literals
        genFn = """

        @PIN(2) = @PIN(0) && @PIN(1);

        """.trimIndent()

        genInit = """



        """.trimIndent()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // PUBLIC FUNCTIONS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    override fun constructInstance(): Block {
        val ins: Block = Block(NAME)

        ins.body = Rectangle()
        ins.symbol = "M -35 -35 l 35 0 c 45 0 45 70 0 70 l -35 0 l 0 -70";

        ins.addPin(Pin.create()
                .withParent(ins)
                .withVariable(Variable(Variable.Type.DISCRETE_SIGNAL))
                .atIndex(0)
                .ofFlow(Pin.Flow.INPUT)
                .onSide(Pin.Side.LEFT)
                .attachedAt(-40, -20))

        ins.addPin(Pin.create()
                .withParent(ins)
                .withVariable(Variable(Variable.Type.DISCRETE_SIGNAL))
                .atIndex(1)
                .ofFlow(Pin.Flow.INPUT)
                .onSide(Pin.Side.LEFT)
                .attachedAt(-40, 20))

        ins.addPin(Pin.create()
                .withParent(ins)
                .withVariable(Variable(Variable.Type.DISCRETE_SIGNAL))
                .atIndex(2)
                .ofFlow(Pin.Flow.OUTPUT)
                .onSide(Pin.Side.RIGHT)
                .attachedAt(40, 0))

        return ins
    }
}